### PR TITLE
Pass inference API token from request to inference service

### PR DESCRIPTION
**Pull request by n0x29a**
This change enables dynamic token propagation from API requests to the inference service. 
Key changes:

- Add InferenceToken type to hold API tokens from requests
- Modify auth middleware to extract and propagate tokens from request headers 
- Update inference service to accept tokens per-request (when possible)
- Remove requirement for token in config (now optional)
- Thread token through query and update operations
- Fall back to config token if request token is not provided

The main goal is to allow per-request inference tracking with the user specific token rather than using a single global token. 

Technical details:
- Inference token is extracted during auth and stored in request extensions
- Token is propagated through all operations that need inference (queries, updates, etc)
- InferenceService.infer() now accepts a token parameter
- Default to config token if no request token is provided

### CODE REVIEW

File: src/common/inference/mod.rs
Comment (agourlay) [inline src/common/inference/mod.rs:17]: Why did you decide for this design instead of having an `Option<InferenceToken>` instead?

Comment (n0x29a) [reply src/common/inference/mod.rs:17]: We discussed with @generall that it could be an option. The reason is to simplify the code. If it's not the best approach, share your ideas to make it better.

File: src/actix/api/query_api.rs
Comment (agourlay) [inline src/actix/api/query_api.rs:66]: I do not think this clone is necessary :)

File: src/actix/api/query_api.rs
Comment (agourlay) [inline src/actix/api/query_api.rs:136]: same here about clone

File: src/actix/api/query_api.rs
Comment (agourlay) [inline src/actix/api/query_api.rs:213]: +1

File: src/common/inference/service.rs
Comment (agourlay) [inline src/common/inference/service.rs:141]: So now we rely solely on this code for the token extraction?

```rust
  let token = inference_token
            .0
            .unwrap_or_else(|| self.config.token.clone().unwrap_or_default());
```

Does the inference service fail with a 4xx error with an empty token?

File: src/common/inference/mod.rs
Comment (agourlay) [inline src/common/inference/mod.rs:20]: Is there nothing to validate about the token String itself?

File: src/tonic/auth.rs
Comment (agourlay) [inline src/tonic/auth.rs:34]: Maybe we can `assert_debug` that the insertion returns None?
As far I understand, there should never be already an inference token on the request.
WDYT?

File: src/common/inference/service.rs
Comment (n0x29a) [reply src/common/inference/service.rs:141]: yes, the inference service fail with a 4xx error with an empty token. it's a change for upcoming release.

File: src/tonic/auth.rs
Comment (n0x29a) [reply src/tonic/auth.rs]: The inference token should be updated per request, because requests could land to the inference service from different instances. Does it make sense?

Comment (timvisee) [reply src/tonic/auth.rs]: In that case, wouldn't each request start with a clean slate? The state we define here is not shared across requests, or is it?

Arnaud suggests a debug assertion as sanity check to ensure we do indeed start with a clean slate for every request. That's a valid suggestion to me, as accidentally mixing tokens across requests will be a nightmare.

File: src/common/inference/mod.rs
Comment (timvisee) [reply src/common/inference/mod.rs:20]: Do we want to allow an empty string?

**Review by timvisee — APPROVED** [review]
Please wait for at least one other approval before merging :smile:

**Review by agourlay — APPROVED** [review]

**Pull request merged**
